// @datorama/akita/esm2015/src/types.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 * @template T
 */
export function HashMap() { }
/**
 * @record
 * @template E, IDType
 */
export function EntityState() { }
if (false) {
    /** @type {?|undefined} */
    EntityState.prototype.entities;
    /** @type {?|undefined} */
    EntityState.prototype.ids;
    /** @type {?|undefined} */
    EntityState.prototype.loading;
    /** @type {?|undefined} */
    EntityState.prototype.error;
    /* Skipping unhandled member: [key: string]: any;*/
}
/**
 * @record
 * @template E
 */
export function Entities() { }
if (false) {
    /** @type {?} */
    Entities.prototype.entities;
    /** @type {?} */
    Entities.prototype.ids;
}
/**
 * @record
 * @template T
 */
export function ActiveState() { }
if (false) {
    /** @type {?} */
    ActiveState.prototype.active;
}
/**
 * @record
 * @template T
 */
export function MultiActiveState() { }
if (false) {
    /** @type {?} */
    MultiActiveState.prototype.active;
}
/**
 * @record
 * @template E
 */
export function SelectOptions() { }
if (false) {
    /** @type {?|undefined} */
    SelectOptions.prototype.asObject;
    /** @type {?|undefined} */
    SelectOptions.prototype.filterBy;
    /** @type {?|undefined} */
    SelectOptions.prototype.limitTo;
}
