// @service-work/is-loading/esm2015/lib/module/directive/is-loading-spinner.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ChangeDetectorRef, ElementRef } from '@angular/core';
export class IsLoadingSpinnerComponent {
    // don't need change detection for this component
    /**
     * @param {?} cdr
     * @param {?} el
     */
    constructor(cdr, el) {
        this.cdr = cdr;
        this.el = el;
        this.cdr.detach();
    }
}
IsLoadingSpinnerComponent.decorators = [
    { type: Component, args: [{
                selector: 'sw-is-loading-spinner',
                template: ``,
                host: { class: 'sw-is-loading-spinner' }
            }] }
];
/** @nocollapse */
IsLoadingSpinnerComponent.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    IsLoadingSpinnerComponent.prototype.cdr;
    /** @type {?} */
    IsLoadingSpinnerComponent.prototype.el;
}
