// @service-work/is-loading/esm5/public_api.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * Public API Surface of is-loading
 *
 * Important!
 * The files must be exported
 * individually without using an `index.ts` file (i.e.
 * `./lib/module/directive`). Otherwise there's some sort of
 * bug with `ng-packagr` or `angular` which prevents the
 * `service-work-is-loading.metadata.json` from being
 * properly built.
 */
export { IsLoadingService } from './lib/is-loading.service';
export { IsLoadingModule } from './lib/module/is-loading.module';
export { IsLoadingDirectiveModule } from './lib/module/directive/is-loading.directive.module';
export { IsLoadingDirective } from './lib/module/directive/is-loading.directive';
export { SW_IS_LOADING_DIRECTIVE_CONFIG } from './lib/module/directive/is-loading.directive.config';
export { IsLoadingPipeModule } from './lib/module/pipe/is-loading.pipe.module';
export { IsLoadingPipe } from './lib/module/pipe/is-loading.pipe';
